File: src/algorithms/arrays/cyclic-sort/find-all-duplicates/sources/find-all-duplicates_test.rs
include!("find-all-duplicates.rs");

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_default_input() {
        let mut result = find_all_duplicates(&[4, 3, 2, 7, 8, 2, 3, 1]);
        result.sort();
        assert_eq!(result, vec![2, 3]);
    }

    #[test]
    fn test_no_duplicates() {
        let result = find_all_duplicates(&[1, 2, 3, 4, 5]);
        assert_eq!(result, vec![]);
    }

    #[test]
    fn test_single_duplicate() {
        let result = find_all_duplicates(&[1, 2, 3, 2]);
        assert_eq!(result, vec![2]);
    }

    #[test]
    fn test_multiple_duplicates() {
        let mut result = find_all_duplicates(&[1, 1, 2, 2, 3, 3]);
        result.sort();
        assert_eq!(result, vec![1, 2, 3]);
    }

    #[test]
    fn test_single_element() {
        let result = find_all_duplicates(&[1]);
        assert_eq!(result, vec![]);
    }

    #[test]
    fn test_empty_array() {
        let result = find_all_duplicates(&[]);
        assert_eq!(result, vec![]);
    }

    #[test]
    fn test_all_appear_twice() {
        let mut result = find_all_duplicates(&[2, 1, 2, 1]);
        result.sort();
        assert_eq!(result, vec![1, 2]);
    }
}

File: src/algorithms/arrays/cyclic-sort/find-all-duplicates/find-all-duplicates_test.go
package findallduplicates

import (
	"reflect"
	"sort"
	"testing"
)

func TestDefaultInput(t *testing.T) {
	result := findAllDuplicates([]int{4, 3, 2, 7, 8, 2, 3, 1})
	sort.Ints(result)
	if !reflect.DeepEqual(result, []int{2, 3}) {
		t.Errorf("Expected [2 3], got %v", result)
	}
}

func TestNoDuplicates(t *testing.T) {
	result := findAllDuplicates([]int{1, 2, 3, 4, 5})
	if len(result) != 0 {
		t.Errorf("Expected empty, got %v", result)
	}
}

func TestSingleDuplicate(t *testing.T) {
	result := findAllDuplicates([]int{1, 2, 3, 2})
	if !reflect.DeepEqual(result, []int{2}) {
		t.Errorf("Expected [2], got %v", result)
	}
}

func TestMultipleDuplicates(t *testing.T) {
	result := findAllDuplicates([]int{1, 1, 2, 2, 3, 3})
	sort.Ints(result)
	if !reflect.DeepEqual(result, []int{1, 2, 3}) {
		t.Errorf("Expected [1 2 3], got %v", result)
	}
}

func TestEmptyArray(t *testing.T) {
	result := findAllDuplicates([]int{})
	if len(result) != 0 {
		t.Errorf("Expected empty, got %v", result)
	}
}

func TestAllAppearTwice(t *testing.T) {
	result := findAllDuplicates([]int{2, 1, 2, 1})
	sort.Ints(result)
	if !reflect.DeepEqual(result, []int{1, 2}) {
		t.Errorf("Expected [1 2], got %v", result)
	}
}
